src/GafferVDB/VDBLevelSetOffset.cpp

```diff
@@ -0,0 +1,142 @@
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "openvdb/openvdb.h"
+#include "openvdb/tools/LevelSetFilter.h"
+
+#include "Gaffer/StringPlug.h"
+
+#include "GafferVDB/VDBObject.h"
+#include "GafferVDB/VDBLevelSetOffset.h"
+
+using namespace std;
+using namespace Imath;
+using namespace IECore;
+using namespace Gaffer;
+using namespace GafferVDB;
+
+
+IE_CORE_DEFINERUNTIMETYPED( VDBLevelSetOffset );
+
+size_t VDBLevelSetOffset::g_firstPlugIndex = 0;
+
+VDBLevelSetOffset::VDBLevelSetOffset( const std::string &name )
+	:	SceneElementProcessor( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	addChild( new StringPlug( "gridName", Plug::In, "density") );
+	addChild( new FloatPlug( "offset", Plug::In, 0.5) );
+}
+
+VDBLevelSetOffset::~VDBLevelSetOffset()
+{
+}
+
+Gaffer::StringPlug *VDBLevelSetOffset::gridNamePlug()
+{
+	return  getChild<StringPlug>( g_firstPlugIndex );
+}
+
+const Gaffer::StringPlug *VDBLevelSetOffset::gridNamePlug() const
+{
+	return  getChild<StringPlug>( g_firstPlugIndex );
+}
+
+Gaffer::FloatPlug *VDBLevelSetOffset::offsetPlug()
+{
+	return  getChild<FloatPlug>( g_firstPlugIndex + 1 );
+}
+
+const Gaffer::FloatPlug *VDBLevelSetOffset::offsetPlug() const
+{
+	return  getChild<FloatPlug>( g_firstPlugIndex + 1 );
+}
+
+void VDBLevelSetOffset::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	SceneElementProcessor::affects( input, outputs );
+
+	if( input == gridNamePlug() || input == offsetPlug() )
+	{
+		outputs.push_back( outPlug()->objectPlug() );
+	}
+}
+
+bool VDBLevelSetOffset::processesObject() const
+{
+	return true;
+}
+
+void VDBLevelSetOffset::hashProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	SceneElementProcessor::hashProcessedObject( path, context, h );
+
+	gridNamePlug()->hash( h );
+	offsetPlug()->hash( h );
+}
+
+IECore::ConstObjectPtr VDBLevelSetOffset::computeProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::ConstObjectPtr inputObject ) const
+{
+	if (ConstVDBObjectPtr vdbObject = runTimeCast<const VDBObject>(inputObject))
+	{
+		std::string gridName = gridNamePlug()->getValue();
+
+		openvdb::GridBase::ConstPtr gridBase = vdbObject->findGrid( gridName );
+
+		if (!gridBase)
+		{
+			return inputObject;
+		}
+
+		openvdb::GridBase::Ptr newGrid = gridBase->deepCopyGrid();
+		openvdb::FloatGrid::Ptr floatGrid = openvdb::GridBase::grid<openvdb::FloatGrid>( newGrid );
+		openvdb::tools::LevelSetFilter<openvdb::FloatGrid> filter ( *floatGrid ) ;
+		filter.offset( offsetPlug()->getValue() );
+		VDBObjectPtr newVDBObject = runTimeCast<VDBObject>( vdbObject->copy() );
+
+		// todo this needs to be a better API
```

Comment:
Seems indicative of the "is it a map? is it a vector?" question I raised earlier.